metadata-ingestion/src/datahub/ingestion/source/data_lake_common/object_store.py

```diff
@@ -0,0 +1,524 @@
+    3. Falls back to specific URI format parsing if needed
+
+    Args:
+        uri: The URI to get the bucket name from
+
+    Returns:
+        The bucket name
+
+    Raises:
+        ValueError: If the URI is not supported by any registered object store
+    """
+    # First try to find the native implementation for this URI
+    object_store = get_object_store_for_uri(uri)
+    if object_store:
+        return object_store.get_bucket_name(uri)
+
+    # If no native implementation, handle specific URI formats directly
+    if uri.startswith("gs://"):
+        return uri[5:].split("/")[0]
+    elif any(uri.startswith(prefix) for prefix in S3ObjectStore.PREFIXES):
+        prefix_length = next(
+            len(prefix) for prefix in S3ObjectStore.PREFIXES if uri.startswith(prefix)
+        )
+        return uri[prefix_length:].split("/")[0]
+    elif uri.startswith(ABSObjectStore.PREFIX):
+        return uri[len(ABSObjectStore.PREFIX) :].split("@")[0]
+
+    raise ValueError(f"Unsupported URI format: {uri}")
+
+
+def get_object_key(uri: str) -> str:
+    """
+    Get the object key from any supported object store URI.
+
+    Args:
+        uri: The URI to get the object key from
+
+    Returns:
+        The object key
+
+    Raises:
+        ValueError: If the URI is not supported by any registered object store
+    """
+    object_store = get_object_store_for_uri(uri)
+    if object_store:
+        return object_store.get_object_key(uri)
+
+    raise ValueError(f"Unsupported URI format: {uri}")
+
+
+class ObjectStoreSourceAdapter:
+    """
+    Adapter for customizing object store source implementations.
+
+    This class provides a way to customize source implementations for different
+    object stores (S3, GCS, etc.) without having to directly modify those classes.
+    Instead, adapters register customizations that are applied to the source instance.
+    """
+
+    @staticmethod
+    def create_s3_path(bucket_name: str, key: str) -> str:
+        """
+        Create a default S3 path.
+
+        Args:
+            bucket_name: The bucket name
+            key: The object key
+
+        Returns:
+            A properly formatted S3 URI
+        """
+        return unquote(f"s3://{bucket_name}/{key}")
+
+    @staticmethod
+    def create_gcs_path(bucket_name: str, key: str) -> str:
+        """
+        Create a default GCS path.
+
+        Args:
+            bucket_name: The bucket name
+            key: The object key
+
+        Returns:
+            A properly formatted GCS URI
+        """
+        return unquote(f"gs://{bucket_name}/{key}")
+
+    @staticmethod
+    def get_s3_external_url(table_data: TableData) -> Optional[str]:
+        """
+        Get the AWS S3 console URL for the given table.
+
+        Args:
+            table_data: Table data containing path information
+
+        Returns:
+            The AWS console URL, or None if not applicable
+        """
+        if not table_data.table_path.startswith("s3://"):
```

Comment:
there are other valid prefixes, right? `s3n`, `s3a`
should this validation be calling `is_uri` instead?